esp-hal/src/timer/mod.rs

```diff
@@ -1,6 +1,231 @@
 //! General-purpose timers.
 
+use fugit::{ExtU64, MicrosDurationU64};
+
 #[cfg(systimer)]
 pub mod systimer;
 #[cfg(any(timg0, timg1))]
 pub mod timg;
+
+/// Timer errors.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// The timer is already active.
+    TimerActive,
+    /// The timer is not currently active.
+    TimerInactive,
+    /// The alarm is not currently active.
+    AlarmInactive,
+}
+
+/// Functionality provided by any timer peripheral.
+pub trait Timer: crate::private::Sealed {
+    /// Start the timer.
+    fn start(&self);
+
+    /// Stop the timer.
+    fn stop(&self);
+
+    /// Reset the timer value to 0.
+    fn reset(&self);
+
+    /// Is the timer running?
+    fn is_running(&self) -> bool;
+
+    /// The current timer value.
+    fn now(&self) -> u64;
+
+    /// Load a target value into the timer.
+    fn load_value(&self, value: MicrosDurationU64);
+
+    /// Enable auto reload of the loaded value.
+    fn enable_auto_reload(&self, auto_reload: bool);
+
+    /// Enable or disable the timer's interrupt.
+    fn enable_interrupt(&self, state: bool);
+
+    /// Clear the timer's interrupt.
+    fn clear_interrupt(&self);
+
+    /// Has the timer triggered?
+    fn is_interrupt_set(&self) -> bool;
+
+    /// FIXME: This is (hopefully?) temporary...
+    fn set_alarm_active(&self, state: bool);
+}
+
+/// A one-shot timer.
+pub struct OneShotTimer<T> {
+    inner: T,
+}
+
+impl<T> OneShotTimer<T>
+where
+    T: Timer,
+{
+    /// Construct a new instance of [`OneShotTimer`].
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+
+    /// Pauses execution for *at least* `ms` milliseconds.
+    pub fn delay_millis(&self, ms: u32) {
+        self.delay((ms as u64).millis());
+    }
+
+    /// Pauses execution for *at least* `us` microseconds.
+    pub fn delay_micros(&self, us: u32) {
+        self.delay((us as u64).micros());
+    }
+
+    /// Pauses execution for *at least* `ns` nanoseconds.
+    pub fn delay_nanos(&self, ns: u32) {
+        self.delay((ns as u64 / 1000).micros())
+    }
+
+    fn delay(&self, us: MicrosDurationU64) {
+        if self.inner.is_running() {
+            self.inner.stop();
+        }
+
+        self.inner.clear_interrupt();
+        self.inner.reset();
+
+        // NOTE: Auto-reload MUST be enabled before loading the value in order
+        //       for `SYSTIMER` to work correctly!
+        self.inner.enable_auto_reload(false);
+        self.inner.load_value(us);
+        self.inner.enable_interrupt(true);
```

Comment:
why are we enabling interrupts here?

Reply:
Why do we even have an `enable_interrupt` function if it doesn't need to be called? I'm  so confused by all this.

Reply:
I think we'll need it later - e.g. for using the abstraction in esp-wifi we'll need a way to enable/disable interrupts and register a handler 🤔 but I think enabling the interrupt should be an explicit step